design/20200326.extensible-certificate-controller.md

```diff
@@ -0,0 +1,388 @@
+          an Event to inform the user.
+        * If the Secret is not labelled as a 'next private key', do nothing and
+          log an Event to inform the user.
+        * If the Secret does not exist:
+            * Generate a new private key according to `spec.privateKey` and
+              create a Secret resource with the given name.
+        * If the secret contains key data that does not match `spec.privateKey`:
+            * Generate a new private key according to `spec.privateKey` and
+              store it in the Secret resource.
+
+* If the `Issuing` condition is `False` or not set:
+    * Delete all owned Secret resources with the `cert-manager.io/next-private-key: "true`
+    * Ensure `status.nextPrivateKeySecretName` is unset - we may want to
+      consider not doing this in case a user has manually specified this field
+      and pointed it at an 'un-owned' Secret. This depends whether we want to
+      support this as a mode of operation.
+
+When creating a 'next private key' Secret resource, the
+`cert-manager.io/next-private-key: "true` annotation is added as well as an
+OwnerReference to the Certificate resource.
+
+Private keys generated by the key manager will be encoded in PKCS#8 format for
+ease of interoperability. The `issuing` controller will encode the resulting
+key-pair into the format requested by the user once the request has been
+completed.
+
+The introduction of this dedicated controller also means we can implement
+private key rotation when a certificate is re-issued/renewed.
+This is a welcome new feature, but in some cases a user may want to pin the
+private key used for a key-pair (as is the default and only supported behaviour
+prior to implementing this design).
+To continue to enable this, the `spec.privateKey.rotationPolicy` field
+controls how the next private key should be sourced. It supports two values:
+
+* `Always`: a new private key will be generated on every re-issuance.
+  This includes renewals as well as changes to the `spec.privateKey` and other
+  `spec` fields.
+* `Never`: private keys will never been regenerated and must be provided by the
+  user.
+
+> TODO: We may be better to have `spec.privateKey.pinnedSecretName` instead, to
+> name a Secret that contains the private key to use. We could then require
+> this key to be provided in a specific format.
+> With the `rotationPolicy` policy design, a user must pre-create the Secret
+> resource containing their private key ahead of time, which creates a conflict
+> of ownership and confusion.
+
+If the `spec.privateKey` options change during an issuance, the key will be
+regenerated and the same Secret resource will be reused to store the updated
+private key. Consumers reading this Secret (i.e. the `requestmanager`) must
+compare the public key of the named 'next private key' and if the public key
+does not match with the CertificateRequest being managed, the
+CertificateRequest should be recreated.
```

Comment:
Might be worth emphasising `during an issuance`, took me a moment to interpret but I think it means if it is changed between the `CertificateRequest` being created and being fulfilled?

Initially reading this I was trying to work out how it relates to the rotation policy options above

Reply:
You're correct, this applies during issuance only. The requestmanager does not do anything except clean up old requests if the `Issuing` condition is not present.